zebrad/src/components/inbound/tests/real_peer_set.rs

```diff
@@ -0,0 +1,498 @@
+//! Inbound service tests with a real peer set.
+
+use std::{collections::HashSet, iter, net::SocketAddr, sync::Arc};
+
+use futures::FutureExt;
+use tokio::{sync::oneshot, task::JoinHandle};
+use tower::{
+    buffer::Buffer,
+    builder::ServiceBuilder,
+    util::{BoxCloneService, BoxService},
+    ServiceExt,
+};
+
+use zebra_chain::{
+    block::{self, Block},
+    parameters::Network,
+    transaction::{AuthDigest, Hash as TxHash, UnminedTxId, WtxId},
+};
+use zebra_consensus::{chain::VerifyChainError, error::TransactionError, transaction};
+use zebra_network::{
+    connect_isolated_tcp_direct, Config as NetworkConfig, Request, Response, ResponseStatus,
+    SharedPeerError,
+};
+use zebra_state::Config as StateConfig;
+use zebra_test::mock_service::{MockService, PanicAssertion};
+
+use crate::{
+    components::{
+        inbound::{Inbound, InboundSetupData},
+        mempool::{self, gossip_mempool_transaction_id, Mempool},
+        sync::{self, BlockGossipError, SyncStatus},
+    },
+    BoxError,
+};
+
+use ResponseStatus::*;
+
+/// Check that a network stack with an empty address book only contains the local listener port,
+/// by querying the inbound service via a local TCP connection.
+///
+/// Uses a real Zebra network stack with a local listener address,
+/// and an isolated Zebra inbound TCP connection.
+#[tokio::test]
+async fn inbound_peers_empty_address_book() -> Result<(), crate::BoxError> {
+    let (
+        // real services
+        connected_peer_service,
+        inbound_service,
+        _peer_set,
+        _mempool_service,
+        _state_service,
+        // mocked services
+        _mock_block_verifier,
+        _mock_tx_verifier,
+        // real tasks
+        block_gossip_task_handle,
+        tx_gossip_task_handle,
+        // real open socket addresses
+        listen_addr,
+    ) = setup().await;
+
+    // Use inbound directly
+    let request = inbound_service.clone().oneshot(Request::Peers);
+    let response = request.await;
```

Comment:
Nit: Having two bindings that have paired names but non-paired types seem confusing to me (i.e., one is a `Future` doing a request, the other is an actual `Response` type).

```suggestion
    let response = inbound_service.clone().oneshot(Request::Peers).await;
```

Reply:
I don't think these bindings will cause much confusion, particularly in the tests.

Next time you want to make a large number of similar suggestions, feel free to open a fix PR. (If you decide to do it for these changes, please base it on PR #3505.)